### PR TITLE
NIRC - Orchard Animal Detail Integration

### LINKED ISSUE #327: [NIRC] Orchard LIS Animal Data Integration
[Spec](https://docs.google.com/document/d/1O0zCKicpj5koXPz6UGVSi_tlQsaJHfiMmT46xftGVio/edit)

### Tasks
- [x] Functional Spec @labkey-martyp 
- [x] Impl Plan and Tasks
- [x] Create new module property 
- [x] Create a query for the orchard data 
- [x] Create a Java class for the content creation containing a method for generating the content  
- [x] Create a java trigger helper
- [x] Modify demographics, protocol/project assignment and housing trigger scripts
- [x] Feature verification
- [x] code review 📌

### PR DESCRIPTION
**Pull request by ankurjuneja**
#### Rationale
This PR creates a framework to generate orchard file based on demographics updates.

#### Changes
* new module property
* new java trigger
* new tests

### CODE REVIEW

File: nirc_ehr/src/org/labkey/nirc_ehr/NIRCOrchardFileGenerator.java
Comment (labkey-jeckels) [inline nirc_ehr/src/org/labkey/nirc_ehr/NIRCOrchardFileGenerator.java]: This isn't doing a provider update. Can you refresh the comment? I'm curious why this is a commit task and why it's being done in a separate thread.

Comment (labkey-jeckels) [reply nirc_ehr/src/org/labkey/nirc_ehr/NIRCOrchardFileGenerator.java]: Also, if a single transaction processes more than one animal ID, won't it write out one file per ID, each time stomping over the file it wrote immediately before?

Comment (labkey-martyp) [reply nirc_ehr/src/org/labkey/nirc_ehr/NIRCOrchardFileGenerator.java]: @labkey-jeckels This is going to be kicked off in various triggers and some of the data in the transaction will be needed in the file. So the cleanest way to ensure some yet unspecified data was committed to the db was to do this post commit. And a separate thread just to not hold up the main UI thread waiting to generate the file.

Writing the file and its contents itself will be left up to the external dev here, but @ankurjuneja can we add the animal Id or something to the file name to just show how an update with multiple animals might work. And update the comment.

Comment (labkey-martyp) [reply nirc_ehr/src/org/labkey/nirc_ehr/NIRCOrchardFileGenerator.java]: And thinking about it, we should probably just get a list of ids from the onUpsert triggers and do the final call to this trigger helper in onComplete with a set of unique Ids.

File: nirc_ehr/src/org/labkey/nirc_ehr/NIRCOrchardFileGenerator.java
Comment (labkey-jeckels) [inline nirc_ehr/src/org/labkey/nirc_ehr/NIRCOrchardFileGenerator.java]: Don't we need separators between the fields?

**Review by labkey-jeckels — APPROVED** [review]

File: nirc_ehr/test/src/org.labkey.test/tests.nirc_ehr/NIRC_EHRTest.java
Comment (labkey-tchad) [inline nirc_ehr/test/src/org.labkey.test/tests.nirc_ehr/NIRC_EHRTest.java]: Since we know the file location, there's no need to delay setting the value (will need to create it later though)
```suggestion
    private static final String orchardFileLocation = TestFileUtils.getTestTempDir();
```

File: nirc_ehr/test/src/org.labkey.test/tests.nirc_ehr/NIRC_EHRTest.java
Comment (labkey-tchad) [inline nirc_ehr/test/src/org.labkey.test/tests.nirc_ehr/NIRC_EHRTest.java]: Remove unnecessary navigation
```suggestion
```

**Review by labkey-tchad — CHANGES_REQUESTED** [review]

File: nirc_ehr/test/src/org.labkey.test/tests.nirc_ehr/NIRC_EHRTest.java
Comment (labkey-tchad) [inline nirc_ehr/test/src/org.labkey.test/tests.nirc_ehr/NIRC_EHRTest.java]: There is a  hidden dependency on `setOrchardFileLocation` here. It would be more clear to create the directory here.
```suggestion
        FileUtils.deleteQuietly(orchardFileLocation);
        FileUtils.forceMkdir(orchardFileLocation);
        List<ModuleProperty> props = List.of(
                new ModuleProperty("EHR", "/" + getProjectName(), "EHRCustomModule", "NIRC_EHR"),
                new ModuleProperty("NIRC_EHR", "/", "NIRCOrchardFileLocation", orchardFileLocation)
        );
```
Should also remove `setOrchardFileLocation`.

**Review by labkey-tchad — CHANGES_REQUESTED** [review]
Remove `setOrchardFileLocation` and fix build.

**Review by labkey-tchad — APPROVED** [review]

**Pull request merged**
